Add a guide for contributing to UniFFI bindings (#1655)
matrix-org/matrix-rust-sdk: bindings/CONTRIBUTING.md

## Introduction
**matrix-rust-sdk** leverages [UniFFI](https://mozilla.github.io/uniffi-rs/) to generate bindings for host languages (eg. Swift and Kotlin).

The process requires to developers to write a `.udl` (UniFFI Definition Language) file, where Rust definitions to be exposed are collected.

Rust code related with bindings live in the [matrix-rust-sdk/bindings](https://github.com/matrix-org/matrix-rust-sdk/tree/main/bindings) folder.

Each Rust crate can expose definitions to UniFFI using two different approaches:

> - Using an `.udl` file. When a crate has one, you find it under the `src` folder (an example is [here](https://github.com/matrix-org/matrix-rust-sdk/blob/main/bindings/matrix-sdk-ffi/src/api.udl)). In this case most of the Rust definitions can be found in the file.

Reviewer: I think "most of the Rust definitions" is going to change fairly soon, so we shouldn't word it like this. The attributes are preferred, where applicable.

- Add UniFFI directivies as Rust attributes. In this case Rust source files (`.rs`) contain attributes related to UniFFI (e.g. `#[uniffi::export]`). 
 

## Expose Rust methods/types to UniFFI

### Check if the API is already on UniFFI

> First of all check if the API/type you are looking for exists on UniFFI already. Most of exposed matrix definitions are collected in the crate [matrix-rust-sdk](https://github.com/matrix-org/matrix-rust-sdk/tree/main/bindings/matrix-sdk-ffi).

Reviewer: ```suggestion
First of all check if the API / type you are looking for exists on UniFFI already. Most of exposed matrix definitions are collected in the crate [matrix-rust-sdk](https://github.com/matrix-org/matrix-rust-sdk/tree/main/bindings/matrix-sdk-ffi).
```

> This crate contains mainly small Rust wrappers around the actual Rust sdk (e.g. the crate [matrix-sdk](https://github.com/matrix-org/matrix-rust-sdk/tree/main/crates/matrix-sdk))

Reviewer: ```suggestion
This crate contains mainly small Rust wrappers around the actual Rust SDK (e.g. the crate [matrix-sdk](https://github.com/matrix-org/matrix-rust-sdk/tree/main/crates/matrix-sdk))
```

> If the API is alreadu on UniFFI, you can probably see it also in the file [matrix-sdk-ffi/src/api.udl](https://github.com/matrix-org/matrix-rust-sdk/blob/main/bindings/matrix-sdk-ffi/src/api.udl) where most of the matrix definitons are declared.

Reviewer 2: ```suggestion
If the API is already on UniFFI, you can probably see it also in the file [matrix-sdk-ffi/src/api.udl](https://github.com/matrix-org/matrix-rust-sdk/blob/main/bindings/matrix-sdk-ffi/src/api.udl) where most of the matrix definitons are declared.
```

Author: Applied -> https://github.com/matrix-org/matrix-rust-sdk/commit/5daf96e655b6eb07a033a1bed97523eb0ad02454

> If the API is on UniFFI already, you can probably see it also in the file [matrix-sdk-ffi/src/api.udl](https://github.com/matrix-org/matrix-rust-sdk/blob/main/bindings/matrix-sdk-ffi/src/api.udl) where most of the matrix definitions are declared.

Reviewer: Same concern here.

### Adding a missing matrix API

> 1. Unless you want to contribute on the crypto side, you problably need to add some code in the [matrix-rust-sdk/bindings/matrix-sdk-ffi](https://github.com/matrix-org/matrix-rust-sdk/tree/main/bindings/matrix-sdk-ffi) crate. After you find the crate you need to undesrtand which file is best to contain the new Rust definition. When exposing new matrix API often (but not always) you need to touch the file [client.rs](https://github.com/matrix-org/matrix-rust-sdk/blob/main/bindings/matrix-sdk-ffi/src/client.rs).

Reviewer 2: ```suggestion
1. Unless you want to contribute on the crypto side, you probably need to add some code in the [matrix-rust-sdk/bindings/matrix-sdk-ffi](https://github.com/matrix-org/matrix-rust-sdk/tree/main/bindings/matrix-sdk-ffi) crate. After you find the crate you need to understand which file is best to contain the new Rust definition. When exposing new matrix API often (but not always) you need to touch the file [client.rs](https://github.com/matrix-org/matrix-rust-sdk/blob/main/bindings/matrix-sdk-ffi/src/client.rs).
```

Author: Applied -> https://github.com/matrix-org/matrix-rust-sdk/commit/5daf96e655b6eb07a033a1bed97523eb0ad02454

> 2. Identify the API to expose in the the target Rust crate (typically in [matrix-sdk](https://github.com/matrix-org/matrix-rust-sdk/tree/main/crates/matrix-sdk). If you can’t find your definition. If you can’t find it, you probably need to touch the actual Rust sdk as well. In this case you typically just need to write some around [ruma](https://github.com/ruma/ruma) (a rust sdk’s dependency) which already implements most of the matrix protocol. 

Reviewer: "If you can’t find your definition." isn't a complete sentence, was it partially deleted by accident?

Author: Yeah, leftover. 
Addressed here: https://github.com/matrix-org/matrix-rust-sdk/pull/1655/commits/0d43a102540aef8b7c9f06d5900ceb0cd5ed464c

3. When you got the Rust code you need to expose to UniFFI, just write a small Rust wrapper in the related UniFFI crate (most of the time is **matrix-sdk-ffi**).

4. When your new definition is ready, remember to expose it in the related `.udl` file (or add an appropriate UniFFI export Rust attribute). For the **matrix-sdk-ffi** crate the definition file is [api.udl](https://github.com/matrix-org/matrix-rust-sdk/blob/main/bindings/matrix-sdk-ffi/src/api.udl). Remember: the language inside a `.udl` file isn’t Rust. To learn more about how map Rust into UDL read [here](https://mozilla.github.io/uniffi-rs/udl_file_spec.html).

## FAQ

> **Q**: I wrote my Rust code and exposed it to UniFFI. How can I check if the compiler is happy?<br/>**A**: Run `cargo build` in the crate you touched (e.g. matrix-sdk-ffi). The compiler will complain if the Rust code and/or the `.udl` is wrong.

Reviewer: ```suggestion
**Q**: I wrote my Rust code and exposed it to UniFFI. How can I check if the compiler is happy?\
**A**: Run `cargo build` in the crate you touched (e.g. matrix-sdk-ffi). The compiler will complain if the Rust code and/or the `.udl` is wrong.
```

> **Q**: The compiler is happy with my code but the CI is failing on GitHub. How can I fix it?<br/>**A**: The CI may fail for different reasons, you need to have a look on the failing GitHub workflow. One common reason though, is that the linter ([Clippy](https://github.com/rust-lang/rust-clippy)) isn’t happy with your code. If this is the case, you can run `cargo clippy` in the crate you touched to see what’s wrong.

Reviewer: ```suggestion
**Q**: The compiler is happy with my code but the CI is failing on GitHub. How can I fix it?\
**A**: The CI may fail for different reasons, you need to have a look on the failing GitHub workflow. One common reason though, is that the linter ([Clippy](https://github.com/rust-lang/rust-clippy)) isn’t happy with your code. If this is the case, you can run `cargo clippy` in the crate you touched to see what’s wrong.
```

3. After you got (by finding or writing) the required Rust code, you need to expose to UniFFI. To do that just write a small Rust wrapper in the related UniFFI crate (most of the time is **matrix-sdk-ffi**) you found on step 1.

4. When your new (wrapping) Rust definition is ready, remember to expose it to UniFFI.
It’s best to do it using UniFFI Rust attributes (e.g. `#[uniffi::export]`). Otherwise add the new definition in the crate’s `.udl` file. For the **matrix-sdk-ffi** crate the definition file is [api.udl](https://github.com/matrix-org/matrix-rust-sdk/blob/main/bindings/matrix-sdk-ffi/src/api.udl). **Remember**: the language inside a `.udl` file isn’t Rust. To learn more about how map Rust into UDL read [here](https://mozilla.github.io/uniffi-rs/udl_file_spec.html)

## FAQ

**Q**: I wrote my Rust code and exposed it to UniFFI. How can I check if the compiler is happy?\

> **A**: Run `cargo build` in the crate you touched (e.g. matrix-sdk-ffi). The compiler will complain if the Rust code and/or the `.udl` is wrong.

Reviewer 3: This might be my lack of understanding, but does `cargo build` actually do anything with the UDL file? I thought you needed to run one of the xtasks to invoke uniffi?

Reviewer: There's two things that are generated from the UDL file:

- Rust scaffolding code (as part of `cargo build`)
- Foreign language bindings (previously upstream uniffi-bindgen binary, now integrated in xtask or our own uniffi-bindgen)

Reviewer 3: Right, that makes sense 👍
